ddl/table.go

```diff
@@ -356,6 +356,69 @@ func (d *ddl) onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ e
 	return ver, nil
 }
 
+// TODO: If may have the issue when two clients concurrently add partitions to a table.
+func (d *ddl) onAddTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	partitionInfo := &model.PartitionInfo{}
+	err := job.DecodeArgs(&partitionInfo)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	originalState := partitionInfo.State
+	switch partitionInfo.State {
+	case model.StateNone:
+		// none -> delete only
+		job.SchemaState = model.StateDeleteOnly
+		partitionInfo.State = model.StateDeleteOnly
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != partitionInfo.State)
+	case model.StateDeleteOnly:
+		// delete only -> write only
+		job.SchemaState = model.StateWriteOnly
+		partitionInfo.State = model.StateWriteOnly
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != partitionInfo.State)
+	case model.StateWriteOnly:
+		// write only -> reorganization
+		job.SchemaState = model.StateWriteReorganization
+		partitionInfo.State = model.StateWriteReorganization
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != partitionInfo.State)
+	case model.StateWriteReorganization:
+		// reorganization -> public
+		tblInfo = updatePartitionInfo(partitionInfo, tblInfo)
+		partitionInfo.State = model.StatePublic
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != partitionInfo.State)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// Finish this job.
+		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
+	default:
+		err = ErrInvalidColumnState.Gen("invalid partition state %v", partitionInfo.State)
+	}
+	return ver, nil
+}
+
+func updatePartitionInfo(partitionInfo *model.PartitionInfo, tblInfo *model.TableInfo) *model.TableInfo {
```

Comment:
Since tblInfo is a pointer,
it seems that we do not need to return *model.TableInfo?